### PR TITLE
Develop database schema, establish connection to database and perform read/writes.

### PR DESCRIPTION
**Pull request by MellowYarker**
By developing the schema, we will have a better idea of what data is important. Currently, the database tables and their columns store information that the backend cannot provide, so in future PRs we will extend the current data structures to enable these features. Some examples include order status, time placed/updated.

In this PR, I plan to have the backend read all the records from the database on start up, and then update/write to the database when orders are submitted, filled, and cancelled.

### CODE REVIEW

**Comment by cacheburner** [comment]
Quickly going to summarize what has been modified, and what I want to fix.

### New Stuff
The following functions all read from the cache first, then the database second:
- account `authentication`
- account `get`
- account `get_mut`
- account `print_user`

I'm still figuring out how I want to connect to the database. Is it better to connect in each function, or pass a connection around as a function argument?

Also, on program start up, we read the users from the database, then update all their pending orders. I think that these things should happen in one dedicated function that has access to private account functions like `fetch_account_pending_orders`.

Furthermore, on start up, we fill the markets with all their pending orders. This will eventually be shrunk down to decrease RAM usage. I also want to avoid reading in users unless they're active.

### To Be Fixed
In most of the database accessing functions mentioned earlier, we don't use our in-memory cache very well... This is because of Rust's borrow checker, it isn't smart enough to understand this:
```
match self.get(&username) {           // <------- immutable reference (*)
    // Cache hit, return reference to account
    Some(account) => return Ok(account),
    // Cache miss, check database
    None => {
        // read from DB then insert to cache
        self.insert(username, account); // <--- rust complains that this is a mutable reference after (*), won't compile!
    }
}
```

So we end up doing a second look up in the cache... wasteful!

File: src/account.rs [hunk 1]
```diff
@@ -171,30 +347,68 @@ impl Users {
...
+                    account.fetch_account_pending_orders(conn);
+                    self.cache_user(account);
+                    // self.id_map.insert(account.id.unwrap(), account.username.clone());
+                    // self.users.insert(account.username.clone(), account);
+                },
+                Err(e) => return Err(e)
+            }
+        }
+
+        // TODO: we call get twice if it was a cache hit.
```
Comment (cacheburner) [inline src/account.rs:272]: Lets make sure we actually resolve this problem in a Rust way. It comes up several times and there's no point in having a cache if we use it like a stupid idiot.

**Comment by cacheburner** [comment]
At this point, I think that the goal of this branch has been completed. The entire database has been integrated with the program, we're just missing chronistic information like execution time, update time, etc.

The unfortunate result of persistence is our performance took a turn for the worst. The time the program spends on the CPU hasn't increased too drastically, but we're communicating with disk (the database) way too often. This is going to have to be addressed in another branch, caching will save the day.

I may tidy up some *todo's* here and there before merging, although some of them require large changes (ex. dropping input references to username, since moving strings is more expensive than moving u32's).

**Review by MellowYarker — COMMENTED** [review]
As mentioned later in the PR, this will be fixed once we move to IDs rather than using a hybrid approach of IDs and usernames. We have to store &mut self rather than &mut `property_of_self, which is what is causing the issue. Essentially, this will be resolved when the parser gets better, and this will happen when the frontend is less of a hassle to deal with (saves user details and sends ID instead of username).

**Pull request merged**